### PR TITLE
Sdboot setmode v2

### PR DESCRIPTION
**Pull request by hdante**
Changes since v1:

- Fixed style
- Added missing code to bootspec.c
- Rename configuration to "console-mode"
- Fixed and rewritten commit messages
- Squashed into 3 commits

Not implemented:
- Still using Atoi (copy/paste code)
- Still missing runtime mode switching

### CODE REVIEW

**Comment by hdante** [comment]
I got pull request number 8086

File: src/boot/efi/console.c [hunk 1]
```diff
@@ -148,9 +151,66 @@ static EFI_STATUS change_mode(UINTN mode) {
...
+
+        return text_area;
+}
+
 static EFI_STATUS mode_auto(UINTN *mode) {
-        /* Mode number 2 is first non standard mode, which is provided by
-         * the device manufacturer, so it should be a good mode.
+        const UINT32 HORIZONTAL_MAX_OK = 1920;
+        const UINT32 VERTICAL_MAX_OK = 1080;
+        const UINT64 VIEWPORT_RATIO = 10;
```
Comment (keszybz) [inline src/boot/efi/console.c]: We'd usually make those `#defines`, but it doesn't matter much.

**Comment by keszybz** [comment]
Works nicely, thanks.

I only wonder if the size calculation is not too conservative. On my X1C Lenovo mode 2 looks like this [https://in.waw.pl/~zbyszek/fedora/x1c-mode2.jpg], and the default is [https://in.waw.pl/~zbyszek/fedora/x1c-modedefault.jpg]. I'd say that even mode 2 is a bit on the small side, and the default is definitely too small.

**Comment by ntrp** [comment]
I tried the functionality but it doesn't work for me.. I tried all the possible options max, auto and various numbers on a dell XPS 9550 with a 4k monitor. Is there a way to debug it and see why nothing different happens whatever I set?

**Comment by hdante** [comment]
@ntrp No, but if there's a problem switching mode, the following message appears:

Error switching console mode to %ld: %r.\r

No message means successful mode change (or you're not using the up-to-date systemd-boot).

On the other hand, you could test all modes using the standard EFI Shell. Download it, for example, from:

https://github.com/tianocore/edk2/blob/master/EdkShellBinPkg/MinimumShell/X64/Shell.efi

https://www.tianocore.org/

**Comment by ntrp** [comment]
I tried to access the UEFI Shell and I have 4 modes available but setting the "change-mode" option doesn't work. I don't have any errors while running bootctl.
Anyway what is the best method to revert to "systemd" from "systemd-git"? At the moment I get "netctl: removing systemd-git breaks dependency 'openresolv'".

**Comment by hdante** [comment]
@ntrp What happens exactly ? If there's an error changing mode, systemd-boot must show a message.

**Comment by ntrp** [comment]
There is no error, I just see the default size list of boot entries. Maybe I missed it.. Anyway I reverted to the standard systemd package at the moment so I guess I will try when it gets into the official repo.

**Comment by fabwu** [comment]
@hdante I'm on a Surface Book 2 and the font is still barely readable. I tried all supported console-modes in EFI shell but changing the mode makes just the shell window smaller but doesn't increase the font size.

Do you have any clue how I can scale the shell window up so it takes up the whole screen and the font size gets increased?

**Comment by hdante** [comment]
> Do you have any clue how I can scale the shell window up so it takes up the whole screen and the font size gets increased?

Not with that code. It uses the trivial text console API and, except for the console mode number, has no knowledge of screen dimensions, fonts, scaling or graphics.

That would require taking another step, either in the direction of configuring EFI fonts or using graphics modes.

**Comment by qdlmcfresh** [comment]
> @hdante I'm on a Surface Book 2 and the font is still barely readable. I tried all supported console-modes in EFI shell but changing the mode makes just the shell window smaller but doesn't increase the font size.
> 
> Do you have any clue how I can scale the shell window up so it takes up the whole screen and the font size gets increased?

Did you find a way to fix this? I have the same problem on my sb2

**Comment by qdlmcfresh** [comment]
> @hdante I'm on a Surface Book 2 and the font is still barely readable. I tried all supported console-modes in EFI shell but changing the mode makes just the shell window smaller but doesn't increase the font size.
> 
> Do you have any clue how I can scale the shell window up so it takes up the whole screen and the font size gets increased?

@fabwu did you find a way to fix this? i have the same problem on my sb2

**Comment by fabwu** [comment]
> @fabwu did you find a way to fix this? i have the same problem on my sb2

No I didn’t. I switched to a MacBook in the meantime so I can’t really help you.

**Pull request merged**